intranet/static/js/bus-shared.js

```diff
@@ -0,0 +1,102 @@
+export function getSocket(base_url, location, document, window, time) {
+    const protocol = (location.protocol.indexOf('s') > -1) ? 'wss' : 'ws';
+    let socket;
+    if (base_url !== '') {
+        socket = new ReconnectingWebSocket(`${protocol}://${base_url}/bus/`);
+    } else {
+        socket = new ReconnectingWebSocket(`${websocketProtocol}://${websocketHost}/bus/`);
+    }
+    socket.automaticOpen = true;
+    socket.reconnectInterval = 2000;
+    socket.maxReconnectInterval = 10000;
+    socket.reconnectDecay = 1.25;
+    socket.timeoutInterval = 5000;
+    socket.maxReconnectAttempts = null;
+
+    let disconnected = false;
+    let disconnected_msg = null;
+    let keepAliveTimeoutId = null;
+
+    socket.onopen = () => {
+        if(keepAliveTimeoutId != null) {
+            clearTimeout(keepAliveTimeoutId);
+            keepAliveTimeoutId = null;
+        }
+
+        if (disconnected_msg) {
+            disconnected_msg.update({
+                message: 'Connection Restored',
+                type: 'success',
+                hideAfter: 3
+            });
+        }
+    };
+
+    socket.onmessage = (event) => {
+        if(keepAliveTimeoutId != null) {
+            clearTimeout(keepAliveTimeoutId);
+            keepAliveTimeoutId = null;
+        }
+
+        var data = JSON.parse(event.data)
+
+        // Don't try and handle keepalives -- when handled, they effectively clear the bus board
+        if(data.type === "keepalive-response") {
+            return;
+        }
+
+        if (time === 'afternoon') {
```

Comment:
I think the `time` variable can be removed.

Instead, you could check for the elements that each time of day uses. If the `#arrived` element exists, you populate the morning fields. If the `#map` element exists, you send the afternoon update event.

This may seem minor, but it would allow you to do some interesting things. For example, you can render both the morning and afternoon portions on the signage page (in separate `<div>`s) and then have some JS to show/hide them based on the time. Then the page never has to refresh; it just updates dynamically.

You could also apply this to the main bus page. I know I said earlier that I'd rather have the morning and afternoon buses on separate pages, but I'd be happy with a single page if it automatically switches over so the user doesn't have to intervene.

Reply:
That makes sense for signage, but I'd like to leave the main bus page how it is because:

1. It allows Security to access whatever page they need, independent of the time. I know the websocket currently doesn't make changes to the buses outside the time frame for that page, but I could see use cases where they might ask for that restriction to be removed.
2. Like you pointed out earlier, it makes testing much, much easier in the future.
3. I think separating the code makes it easier to understand and maintain.
4. Having morning/afternoon code on the same page and switching between them with Javascript feels like it defeats the point of a django based system.

In essence, I agree on signage, but I think leaving the main bus page as it is makes it more robust. Thoughts?

Reply:
Hmm. @theo-o what do you think about this?

Reply:
I can see reasonable arguments for both sides of this proposal. I think it ultimately comes down to whether the benefits of having morning and afternoon portions broken out into different pages outweigh the benefits of having the portions exist on the same page.

In my opinion, @Laur04's proposal is the best (broken out on `/bus` and together on `/signage/display/[SIGNAGE]`) because:
- Bus signage is view-only, while regular bus is both read and write
- As a result, there really isn't a need for separate pages for morning and afternoon on bus signage
- Bus signage is only really applicable in the mornings and afternoons, while people may want to see both morning and afternoon on regular bus
- Security may want to see how both the morning and afternoon look like by just visiting different pages instead of waiting for the turnover

Reply:
Sounds good to me.